theano/sandbox/cuda/dnn.py

```diff
@@ -2184,8 +2184,406 @@ def method(self):
+
+    Parameters
+    ----------
+    mode
+        Normalization is performed per-activation for 'per-activation' (in this mode bnBias and bnScale tensor
+        dimensions are 1xCxHxW for 4d tensor) or Normalization is performed over N+spatial dimensions for 'spatial'
+        (in this mode bnBias, bnScale tensor dimensions are 1xCx1x1).
+    epsilon
+        Epsilon value used in the batch normalization formula. It is 1e-4 by default. Minimum allowed value
+        for cuDNN v.4 is 1e-5.
+
+    """
+
+    direction = 'inference'
+    input_descs = [('bn_input', 0), ('scale_bias_mean', 1)]
+
+    def make_node(self, x, bn_scale, bn_bias, estimated_mean, estimated_variance):
+        x = as_cuda_ndarray_variable(x)
+        bn_scale = as_cuda_ndarray_variable(bn_scale)
+        bn_bias = as_cuda_ndarray_variable(bn_bias)
+        estimated_mean = as_cuda_ndarray_variable(estimated_mean)
+        estimated_variance = as_cuda_ndarray_variable(estimated_variance)
+        assert x.ndim == 4
+        assert bn_scale.ndim == 4
+        assert bn_bias.ndim == 4
+        assert estimated_mean.ndim == 4
+        assert estimated_variance.ndim == 4
+        return Apply(self, [x, bn_scale, bn_bias, estimated_mean, estimated_variance],
+                     [x.type()])
+
+    def method(self):
+        return """
+#ifndef CUDNN_VERSION
+err%(name)s = cudnnBatchNormalizationForwardInference(
+  _handle,
+  mode%(name)s,
+  bn_input_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins0)s),
+  bn_output_%(name)s,
+  CudaNdarray_DEV_DATA(%(outs0)s),
+  scale_bias_mean_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins1)s),
+  CudaNdarray_DEV_DATA(%(ins2)s),
+  CudaNdarray_DEV_DATA(%(ins3)s),
+  CudaNdarray_DEV_DATA(%(ins4)s),
+  epsilon%(name)s
+);
+#else
+{
+const float alpha = 1.;
+const float beta = 0.;
+err%(name)s = cudnnBatchNormalizationForwardInference(
+  _handle,
+  mode%(name)s,
+  (void*) &alpha,
+  (void*) &beta,
+  bn_input_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins0)s),
+  bn_output_%(name)s,
+  CudaNdarray_DEV_DATA(%(outs0)s),
+  scale_bias_mean_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins1)s),
+  CudaNdarray_DEV_DATA(%(ins2)s),
+  CudaNdarray_DEV_DATA(%(ins3)s),
+  CudaNdarray_DEV_DATA(%(ins4)s),
+  epsilon%(name)s
+);
+}
+#endif
+"""
+
+
+class GpuDnnBatchNorm(GpuDnnBatchNormBase):
+    """
+    Op for the cuDNN BatchNormalizationForwardTraining
+
+    Parameters
+    ----------
+    mode
+        Normalization is performed per-activation for 'per-activation' (in this mode bnBias and bnScale tensor
+        dimensions are 1xCxHxW for 4d tensor) or Normalization is performed over N+spatial dimensions for 'spatial'
+        (in this mode bnBias, bnScale tensor dimensions are 1xCx1x1).
+    epsilon
+        Epsilon value used in the batch normalization formula. It is 1e-4 by default. Minimum allowed value
+        for cuDNN v.4 is 1e-5.
+
+    """
+
+    direction = 'forward'
+    input_descs = [('bn_input', 0), ('scale_bias_mean', 1)]
+    destroy_map = {1: [3], 2: [4]}
```

Comment:
> Normally, we don't do that. This is because if the user start to decide which op are inplace, it could create cycle in the computation graph without knowing. So this is better left to the optimizer that will do it and make sure to not introduce cycle.

This would require the batch norm Op to be callable like:
```python
output, new_statistics = GpuDnnBatchNorm(inplace=False)(input, old_statistics)
```
However, the implementation in cuDNN does not support updating the statistics (running averages of mean and invstddev) out-of-place at all. It is always destructive.

What would you suggest then? It would be possible to create a non-destructive version by actually copying the inputs before calling the cuDNN implementation, is this the way to go?

Furthermore, if there is a single inplace parameter, how do you know if it applies to the batchnorm statistics or the actual output? Both could be inplace or out-of-place independently.